docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,381 @@
+# API Types package layout
+
+We have written tooling around api types and will write more. This document
+seeks to prescribe an organizational layout for types packages and their
+associated generated files.
+
+Explicit goals are:
+* Organizing our chaos.
+* Make the toolchain easy to understand and use.
+* Reduce the package import graph to a minimum for any given use case.
+* Make it possible to set up a Kubernetes-style types directory in an arbitrary
+  location; decouple our tools and github repository.
+* Coordinate among those who will or might write autogeneration.
+* Where does validation code go? How is it called?
```

Comment:
I think I'm gonna say that construction is outside the scope of this, and we should figure that out when we feel like doing away with the "internal" versions of objects.